hcl2template/types.datasource.go

```diff
@@ -0,0 +1,154 @@
+	}
+}
+
+func (ds *Datasources) Values() (map[string]cty.Value, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	res := map[string]cty.Value{}
+	valuesMap := map[string]map[string]cty.Value{}
+
+	for ref, datasource := range *ds {
+		if datasource.value == (cty.Value{}) {
+			diags = append(diags, &hcl.Diagnostic{
+				Summary:  fmt.Sprintf("empty value"),
+				Subject:  &datasource.block.DefRange,
+				Severity: hcl.DiagError,
+			})
+			continue
+		}
+		inner := valuesMap[ref.Type]
+		if inner == nil {
+			inner = map[string]cty.Value{}
+		}
+		inner[ref.Name] = datasource.value
+		res[ref.Type] = cty.MapVal(inner)
+
+		// Keeps values of different datasources from same type
+		valuesMap[ref.Type] = inner
+	}
+
+	return res, diags
+}
+
+type DatasourceRef struct {
+	Type string
+	Name string
+}
+
+// the 'addition' field makes of ref a different entry in the data sources map, so
+// Ref is here to make sure only one is returned.
+func (r *DatasourceRef) Ref() DatasourceRef {
+	return DatasourceRef{
+		Type: r.Type,
+		Name: r.Name,
+	}
+}
+
+func (cfg *PackerConfig) startDatasource(dataSourceStore packer.DatasourceStore, ref DatasourceRef) (packersdk.Datasource, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	block := cfg.Datasources[ref].block
+
+	datasource, err := dataSourceStore.Start(ref.Type)
+	if err != nil {
+		diags = append(diags, &hcl.Diagnostic{
+			Summary:  err.Error(),
+			Subject:  &block.DefRange,
+			Severity: hcl.DiagError,
+		})
+	}
+	if datasource == nil {
+		diags = append(diags, &hcl.Diagnostic{
+			Summary:  fmt.Sprintf("failed to start datasource plugin %q.%q", ref.Type, ref.Name),
+			Subject:  &block.DefRange,
+			Severity: hcl.DiagError,
+		})
+	}
+	body := block.Body
+	decoded, moreDiags := decodeHCL2Spec(body, cfg.EvalContext(nil), datasource)
+	diags = append(diags, moreDiags...)
+	if moreDiags.HasErrors() {
+		return nil, diags
+	}
+
+	if err := datasource.Configure(decoded); err != nil {
+		diags = append(diags, &hcl.Diagnostic{
+			Summary:  err.Error(),
+			Subject:  &block.DefRange,
+			Severity: hcl.DiagError,
+		})
+	}
+	return datasource, diags
+}
+
+func (p *Parser) decodeDataBlock(block *hcl.Block) (*Datasource, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	r := &Datasource{
+		Type:  block.Labels[0],
+		Name:  block.Labels[1],
+		block: block,
+	}
+
+	if p.DatasourceSchemas == nil {
+		diags = append(diags, &hcl.Diagnostic{
+			Summary:  "Unknown " + dataSourceLabel + " type " + r.Type,
+			Subject:  block.LabelRanges[0].Ptr(),
+			Detail:   fmt.Sprintf("packer does not currently know any data source."),
+			Severity: hcl.DiagError,
+		})
+		return r, diags
+	}
+
+	if !p.DatasourceSchemas.Has(r.Type) {
+		diags = append(diags, &hcl.Diagnostic{
+			Summary:  "Unknown " + dataSourceLabel + " type " + r.Type,
+			Subject:  block.LabelRanges[0].Ptr(),
+			Detail:   fmt.Sprintf("known data sources: %v", p.DatasourceSchemas.List()),
+			Severity: hcl.DiagError,
+		})
+		return r, diags
+	}
```

Comment:
Could you please move this check to `startDatasource`, for #10304 we will check the existence of plugins only in Parse with this PR.  ( I can also do it there if you want ! )

Reply:
✅